tests/ballerina-unit-test/src/test/java/org/ballerinalang/test/nativeimpl/functions/io/socket/ServerSocketTest.java

```diff
@@ -46,38 +47,72 @@
 public class ServerSocketTest {
 
     private static final Logger log = LoggerFactory.getLogger(ServerSocketTest.class);
-    private CompileResult serverBal;
+    private CompileResult normalServer;
+    private final String welcomeMsg = "Hello Ballerina\n";
+    private boolean isConnected = false;
 
     @BeforeClass
     public void setup() {
-        serverBal = BCompileUtil.compileAndSetup("test-src/io/server_socket_io.bal");
+        normalServer = BCompileUtil.compileAndSetup("test-src/io/server_socket_io.bal");
     }
 
     @Test(description = "Check server socket accept functionality.")
-    public void testSeverSocketAccept() {
+    public void testSeverSocketAccept() throws InterruptedException {
         int port = ThreadLocalRandom.current().nextInt(47000, 51000);
-        String welcomeMsg = "Hello Ballerina\n";
         ExecutorService executor = Executors.newSingleThreadExecutor();
         executor.execute(() -> {
             try {
                 BValue[] args = { new BInteger(port), new BString(welcomeMsg) };
-                BRunUtil.invokeStateful(serverBal, "startServerSocket", args);
+                BRunUtil.invokeStateful(normalServer, "startServerSocket", args);
             } catch (Throwable e) {
                 log.error(e.getMessage(), e);
             }
         });
+        Thread.sleep(2500);
+        connectClient(port, 1000, normalServer);
+        executor.shutdownNow();
+    }
+
+    @Test(description = "Check server socket accept functionality.",
+          dependsOnMethods = "testSeverSocketAccept")
+    public void testSeverSocketDelayiedAccept() {
+        CompileResult delayedStartServer = BCompileUtil
+                .compileAndSetup("test-src/io/server_socket_io_delayed_accept.bal");
+        SelectorManager.start();
+        int port = ThreadLocalRandom.current().nextInt(47000, 51000);
+        ExecutorService executor = Executors.newSingleThreadExecutor();
+        isConnected = false;
+        executor.execute(() -> {
+            try {
+                BRunUtil.invokeStateful(delayedStartServer, "initServer", new BValue[] { new BInteger(port) });
+                for (int i = 0; i < 10; i++) {
+                    if (isConnected) {
+                        break;
+                    } else {
+                        Thread.sleep(1000);
+                    }
+                }
+                BRunUtil.invokeStateful(delayedStartServer, "startServerSocket",
+                        new BValue[] { new BString(welcomeMsg) });
+            } catch (Throwable e) {
```

Comment:
We are catching throwable here. Whats the reason for this? Also we are not failing the test case when error/exception is thrown, but just logging it. Is that correct?

Reply:
There can be various exceptions thrown from ballerina code due to syntax or execution error, most of the time without caching throwable, we are unable to get an idea about it. This is just for logging purposes. There will be assert in the below to verify the functionality. That will fail anyway for an incorrect/error situation.